### DISCUSSION TITLE
useMutation ts(2769)

### DISCUSSION THREAD

**Discussion by lorenzoMrt** [TanStack/query#4872] [Q&A] [answered]
For the record I've already found some questions related to this error: 
- [https://github.com/TanStack/query/issues/1962](https://github.com/TanStack/query/issues/1962)
- [https://stackoverflow.com/questions/74064124/react-query-usemutation-no-overload-matches-this-call](https://stackoverflow.com/questions/74064124/react-query-usemutation-no-overload-matches-this-call)
- [https://github.com/TanStack/query/discussions/2350](https://github.com/TanStack/query/discussions/2350)

Although I tested their solutions, none worked for me. So here I am.
Here is the snippet that is giving me the error:
```typescript
function uploadFiles(files: File[], folderPath: string[]) {
  const uploadPromises = files.map((file: File) => {
    const path = folderPath.slice(1, folderPath.length).join('/')
    const fullPath = path === '' ? `${file.name}` : `${path}/${file.name}`

    return uploadFile(fullPath, file)
  })
  return Promise.all(uploadPromises)
    .then(() => {
      toast('Files uploaded!', {
        ...toastDefaultConfig,
        type: toast.TYPE.SUCCESS,
        transition: Flip,
      })
      toast.clearWaitingQueue()
      toast.dismiss()
      reload()
    })
    .catch((e) => getLogger().error(e))
}
const uploadFileMutation = useMutation<unknown, unknown, {files: File[]; folderPath: string[]}>(
    ({files, folderPath}) => uploadFiles(files, folderPath),
    {
      onSettled: () => {
        setTimeout(() => queryClient.invalidateQueries(['files']), 5000)
      },
    }
  )

function handleOnDelete(fileNames: string[]) {
  return Promise.all<Promise<string>[]>(
    fileNames.filter((v) => v.length > 0).map((fileName) => removeFile(fileName))
  )
}
const deleteFileMutation = useMutation<unknown, unknown, string[]>(
  {
    mutationFn: (fileNames: string[]): Promise<string[]> => handleOnDelete(fileNames), //No overload matches this call ts(2769)
    onSuccess: (fileNames: string[]) => {
      queryClient.cancelQueries(['files'])
      const previousFiles = queryClient.getQueryData(['files'])
      queryClient.setQueryData(['files'], (old: any) => {
        return old.filter((file: any) => !fileNames.includes(file.name))
      })
      const rollback: () => void = () => queryClient.setQueryData(['files'], previousFiles)
      return {rollback}
    },
    onError: (err: any, variables: any, context: {rollback: () => void}) => {
      context.rollback()
    },
    onSettled: () => {
      toast('Files deleted', toastDefaultConfig)
      resetSelection()
      reload()
    },
  }
)
```

As you can see, I already annotated the function with a concrete return type as suggested on #1962
What puzzles me the most is that I've already a useMutation working (uploadFileMutation) and I have no ts issues

**Additional context**
typescript: 4.9.4
@tanstack/react-query: 4.22.0

**Comment by louis-young** [comment]
Hi 👋 

Please could you provide a minimal runnable reproduction so we can take a look at this? It's often easier to investigate a minimal runnable reproduction as there's usually less to digest, especially if there are fewer moving parts.

**Reply by lorenzoMrt** [reply]
Sure!

Here it is: [FileTable minimal example](https://codesandbox.io/s/modest-breeze-1wfgmv?file=/src/App.tsx)

P.S: the mentioned error ( ts(2769) ) I've already taken care of (I was declaring onSuccess before onMutate)but, I still have some issues related to typescript. Now, when I try to pass a rollback function to restore the cache after an optimistic update, context is of type unknown. 
Check the codesandbox and you'll see the error on FileTable.tsx
If you could shed some light It would be top 🙌

**Reply by louis-young** [reply]
I've fixed it for you below, provided an explanation and marked this as answered 🙂

**Comment by louis-young** [comment, marked as answer]
Thanks for supplying the reproduction! Here's a [forked fixed version](https://codesandbox.io/s/cranky-hellman-zz0wlf?file=/src/FileTable.tsx).

I fixed it by leveraging type inference. When you supply generics to a function with multiple generics in its signature in TypeScript, any generics that you don't explicitly specify will fall back to their default type (in this case, `unknown`). Inference is usually a great thing. I removed the partially provided generics from the `useMutation` call site, moved the variables argument type for the `mutationFuntion` to be annotated in the mutation function, and checked the context `rollback` function for nullishness before calling it. 

If you're interested in further reading on optimistic updates with TypeScript, please see: https://tkdodo.eu/blog/react-query-and-type-script#optimistic-updates

Hopefully this helps 🙂

**Reply by lorenzoMrt** [reply]
It does help. Thank you! I'll read carefully the article you shared
